### PR TITLE
Miche Population Calculation Fixes

### PR DESCRIPTION
**Pull request by GameDungeon**
**Brief Description of What This PR Does**

Fixes:
- [ ] Auto-Evo prediction incorrect
- [x] Species gain energy from failed Miches Closes:  #5391
- [x] Fix Bug that makes auto-evo think there is iron in every patch

**Related Issues**

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/auto-evo/simulation/MichePopulation.cs [hunk 1]
```diff
@@ -213,15 +213,12 @@ private static void SimulatePatchStep(SimulationConfiguration simulationConfigur
...
         var leafNodes = new List<Miche>();
-        miche.GetLeafNodes(leafNodes, x => x.Occupant != null);
+        miche.GetLeafNodes(leafNodes, x => x.Occupant is MicrobeSpecies);
 
         // TODO: check if energy should be calculated as doubles because the summed numbers can get pretty high that
         // might benefit from extra precision
         var energyDictionary = species.ToDictionary(x => x, _ => 0.0f);
 
-        // This doesn't use miche.SetupScores as this wants scores for all species, even ones that aren't in the
-        // tree yet
```
Comment (hhyyrylainen) [inline src/auto-evo/simulation/MichePopulation.cs:223]: Why was this comment removed?

File: src/auto-evo/simulation/MichePopulation.cs [hunk 2]
```diff
@@ -213,15 +213,12 @@ private static void SimulatePatchStep(SimulationConfiguration simulationConfigur
             return;
 
         var leafNodes = new List<Miche>();
-        miche.GetLeafNodes(leafNodes, x => x.Occupant != null);
+        miche.GetLeafNodes(leafNodes, x => x.Occupant is MicrobeSpecies);
```
Comment (hhyyrylainen) [inline src/auto-evo/simulation/MichePopulation.cs:216]: I'm pretty worried about this change because once we add other species types to auto-evo I reckon this might be forgotten and wreak havoc. What's the reason behind this change?

File: src/auto-evo/simulation/MichePopulation.cs (hunk 1 above)
Comment (GameDungeon) [reply src/auto-evo/simulation/MichePopulation.cs:223]: You removed miche.SetupScores as a function, so I figured we no longer needed to explain why it was not used.

Comment (hhyyrylainen) [reply src/auto-evo/simulation/MichePopulation.cs:223]: Okay that makes sense.

File: src/auto-evo/simulation/MichePopulation.cs (hunk 2 above)
Comment (hhyyrylainen) [reply src/auto-evo/simulation/MichePopulation.cs:216]: Reading the later code I now think I understand this but this should get a pretty big TODO and warning slapped on that this will be majorly wrong once other than microbe species can be fed into auto-evo.

Comment (GameDungeon) [reply src/auto-evo/simulation/MichePopulation.cs:216]: We need to make sure that a later call to the return from this is not null or not a microbe species. I && the microbe species check to this, but the lint wanted me to simplify it. Autoevo is currently littered with checks like this as Multicell species are partially supported. I doubt that this specific check would be missed as there are so many other like it I would probably just crtl + f "is MicrobeSpecies".

Comment (hhyyrylainen) [reply src/auto-evo/simulation/MichePopulation.cs:216]: Could you still add the TODO comment for my piece of mind here?
In most cases in my head it is that Species is usually used but just before microbe specific stuff is done it is then checked, but in this situation the other species types are pruned from the data much earlier.

Comment (GameDungeon) [reply src/auto-evo/simulation/MichePopulation.cs:216]: I can add that TODO

File: src/microbe_stage/systems/ProcessSystem.cs
Comment (GameDungeon) [inline src/microbe_stage/systems/ProcessSystem.cs:1]: That is a cursed way for git to merge that

File: src/auto-evo/steps/MigrateSpecies.cs [hunk 3]
```diff
@@ -99,7 +97,9 @@ public bool RunStep(RunResults results)
                 var moveAmount = (long)random.Next(population * Constants.AUTO_EVO_MINIMUM_MOVE_POPULATION_FRACTION,
                     population * Constants.AUTO_EVO_MAXIMUM_MOVE_POPULATION_FRACTION);
 
-                if (moveAmount > 0 && targetMiche.InsertSpecies(species, patch, null, cache, true, speciesWorkMemory))
+                var insertWorkMemory = new Miche.InsertWorkingMemory(targetMiche);
```
Comment (hhyyrylainen) [inline src/auto-evo/steps/MigrateSpecies.cs:100]: Why was this made a local variable? I'd prefer the field approach as it ensures the list isn't recreated each time in this loop which burns through a bunch of unnecessary memory allocations.

File: src/auto-evo/Miche.cs [hunk 4]
```diff
@@ -246,4 +266,21 @@ public override int GetHashCode()
         return Pressure.GetHashCode() * 131 ^ parentHash * 587 ^
             (Occupant == null ? 17 : Occupant.GetHashCode()) * 5171;
     }
+
+    public class InsertWorkingMemory
```
Comment (hhyyrylainen) [inline src/auto-evo/Miche.cs:270]: Documenting this class and making this reusable for new miches would help a lot in this class' usage.

File: src/auto-evo/Miche.cs [hunk 5]
```diff
@@ -156,26 +177,25 @@ public bool InsertSpecies(Species species, Patch patch, Dictionary<Species, floa
             return true;
         }
 
-        // TODO: somehow avoid this recursive method call memory allocations
-        var newScores = new Dictionary<Species, float>();
+        var newScores = workingMemory.ScoreDictionaries[depth];
```
Comment (hhyyrylainen) [inline src/auto-evo/Miche.cs:180]: Couldn't the cache be made so that this automatically increased the dictionary size if depth is higher? That way the dictionary wouldn't need to be pre-populated making it easier to use? Or could maybe even a stack (or maybe a queue) type be used instead of a dictionary? Or maybe just a plain list which is indexed by the depth? Dictionary lookups are more expensive than list index lookups.

File: src/microbe_stage/systems/ProcessSystem.cs [hunk 6]
```diff
@@ -427,73 +427,76 @@ public static (float Production, float Consumption) CalculateOrganelleATPBalance
...
+        return (processATPProduction, processATPConsumption);
+    }
 
-                foreach (var input in processData.WritableInputs)
+    /// <summary>
+    ///   Calculates the maximum speed a process can run at in a biome based on the environmental compounds.
+    ///   Can be switched between the average, maximum etc. conditions that occur in the span of an in-game day.
+    /// </summary>
+    public static ProcessSpeedInformation CalculateProcessMaximumSpeed(TweakedProcess process,
+        BiomeConditions biome, CompoundAmountType pointInTimeType, bool checkIfInPatch)
```
Comment (hhyyrylainen) [inline src/microbe_stage/systems/ProcessSystem.cs:456]: I think some other name for the parameter is needed. Something like `bool requireInputCompoundsInBiome`?

**Review by hhyyrylainen — APPROVED** [review]
I added some slight code cleanups and verified the self eating is now gone. I did see a random bug with the jumping into the game from the auto-evo tool where my species got 0 energy no matter what I did until I exited the editor and got back to it, but that's probably a totally unrelated bug to these changes.
